packages/harmony/src/components/input/TextInput/TextInput.tsx

```diff
@@ -0,0 +1,178 @@
+  const showMaxlengthWarningColor =
+    maxLength && characterCount >= maxLengthWarningThreshold * maxLength
+
+  // Styles for the root of the input
+  const inputRootStyle = {
+    [styles.default]: size === TextInputSize.DEFAULT,
+    [styles.small]: size === TextInputSize.SMALL,
+    [styles.warning]: warningProp,
+    [styles.error]: error,
+    [styles.focused]: isFocused,
+    [styles.disabled]: disabled,
+    [styles.required]: required
+  }
+
+  // Styles for the input element itself
+  const inputElStyle = {
+    [styles.default]: size === TextInputSize.DEFAULT,
+    [styles.small]: size === TextInputSize.SMALL,
+    [styles.disabled]: disabled
+  }
+
+  const inputRender = (
+    <Flex
+      className={cn(styles.inputAdornmentRow, { [styles.focused]: isFocused })}
+    >
+      {startAdornmentText && shouldShowAdornments ? (
+        <Text variant='label' size='l' color='subdued'>
+          {startAdornmentText}
+        </Text>
+      ) : null}
+      <input
+        onFocus={handleFocus}
+        onBlur={handleBlur}
+        ref={ref}
+        className={cn(styles.inputElement, inputClassName, inputElStyle)}
+        value={value}
+        maxLength={maxLength}
+        disabled={disabled}
+        placeholder={shouldShowPlaceholder ? placeholder : undefined}
+        aria-label={labelText ?? props['aria-label']}
+        aria-required={required}
+        id={id}
+        {...other}
+      />
+      {endAdornmentText && shouldShowAdornments ? (
+        <Text variant='label' size='l' color='subdued'>
+          {endAdornmentText}
+        </Text>
+      ) : null}
+    </Flex>
+  )
+
+  return (
+    <Flex className={cn(styles.root, className)} direction='column'>
+      <label
+        htmlFor={id}
+        className={cn(styles.inputRoot, inputRootClassName, inputRootStyle)}
+      >
+        {StartIcon ? <StartIcon /> : null}
+        <div className={styles.innerRootContainer}>
+          {shouldShowLabel ? (
+            <Flex
+              className={styles.labelRow}
+              direction='row'
+              alignItems='center'
+              justifyContent='space-between'
+              gap='s'
+            >
+              <Text
+                variant='body'
+                tag='span'
+                size='l'
+                className={cn(styles.label, {
+                  [styles.hasValue]: hasValue
+                })}
+              >
+                {labelText}
+              </Text>
+              {shouldShowMaxLengthText ? (
+                <Text
+                  variant='body'
+                  size='xs'
+                  tag='span'
+                  color={showMaxlengthWarningColor ? 'warning' : 'default'}
+                >
+                  {characterCount}/{maxLength}
+                </Text>
+              ) : null}
+            </Flex>
+          ) : null}
+          {inputRender}
+        </div>
+
+        {EndIcon ? <EndIcon /> : null}
+        {children}
```

Comment:
Do we want children to render inside the label? That feels somewhat odd to me as we could explicitly specify a label prop and have the actual label be different

Reply:
It might have been for the helper text? Or icons? I don't remember 🤦‍♂️